Wikipedia/assets/codemirror/codemirror-editTextSelection.js

```diff
@@ -0,0 +1,126 @@
+          return regex
+        }    
+      }
+      return null
+    }
+    const restrictiveRegex = getWorkingRegexWithMostAdjacentWords(this.getRestrictiveRegex.bind(this))
+    if (restrictiveRegex) {
+      return restrictiveRegex
+    }
+    const permissiveRegex = getWorkingRegexWithMostAdjacentWords(this.getPermissiveRegex.bind(this))
+    if (permissiveRegex) {
+      return permissiveRegex
+    }
+    return null
+  }
+
+  // Faster, less error prone - but has more trouble with skipping past some unexpectedly complicated wiki markup, so more prone to not finding any match.
+  getRestrictiveRegex(maxAdjacentWordsToUse) {
+    const atLeastOneNonWordPattern = '\\W+'
+    return this.regexForLocatingSelectedTextWithPatternForSpace(atLeastOneNonWordPattern, maxAdjacentWordsToUse)
+  }
+  
+  // Slower, more error prone - but has less trouble with skipping past some unexpectedly complicated wiki markup, so less prone to not finding any match.
+  getPermissiveRegex(maxAdjacentWordsToUse) {
+    const atLeastOneCharPattern = '.+'
+    return this.regexForLocatingSelectedTextWithPatternForSpace(atLeastOneCharPattern, maxAdjacentWordsToUse)
+  }
+    
+  regexForLocatingSelectedTextWithPatternForSpace(patternForSpace, maxAdjacentWordsToUse) {
+    const replaceSpaceWith = (s, replacement) => s.replace(/\s+/g, replacement)
+
+    // Keep only the last 'maxAdjacentWordsToUse' words of 'textBeforeSelectedText'
+    const shouldKeepWordBeforeSelection = (e, i, a) => (a.length - i - 1) < maxAdjacentWordsToUse
+    // Keep only the first 'maxAdjacentWordsToUse' words of 'textAfterSelectedText'
+    const shouldKeepWordAfterSelection = (e, i) => i < maxAdjacentWordsToUse
+    const wordsBefore = this.textBeforeSelectedText.split(' ').filter(shouldKeepWordBeforeSelection).join(' ')
+    const wordsAfter = this.textAfterSelectedText.split(' ').filter(shouldKeepWordAfterSelection).join(' ')
+
+    const selectedTextPattern = replaceSpaceWith(this.selectedText, patternForSpace)
+    const textBeforeSelectedTextPattern = replaceSpaceWith(wordsBefore, patternForSpace)
+    const textAfterSelectedTextPattern = replaceSpaceWith(wordsAfter, patternForSpace)
+
+    // Attempt to locate wikitext selection based on the non-wikitext context strings above.
+    const atLeastOneNonWordOrOptionalStringPattern = '(?:\\W+|.*)'
+    const pattern = `(${textBeforeSelectedTextPattern.length > 0 ? '.*?' : ''}${textBeforeSelectedTextPattern}${atLeastOneNonWordOrOptionalStringPattern})(${selectedTextPattern})${atLeastOneNonWordOrOptionalStringPattern}${textAfterSelectedTextPattern}`
+    const regex = new RegExp(pattern, 's')
+
+    return regex
+  }
+}
+
+const wikitextRangeForSelectedAndAdjacentText = (selectedAndAdjacentText, wikitext) => {
+  const regex = selectedAndAdjacentText.regexForLocatingSelectedTextInWikitext(wikitext)
+  if (regex === null) {
+    return null
+  }
+  const match = wikitext.match(regex)
+  const matchedWikitextBeforeSelection = match[1]
```

Comment:
should we check if `match` is not null before we try to retrieve values?

Reply:
for sure! fixed!